feat: Prover node

spalladino opened the pull request:
Introduces a new prover node package, which spawns a new prover that follows the proven chain, and exposes a high-level method for proving an unfinalised block. This involves fetching the tx hashes from L1, requesting the individual txs from a tx source (today a remote Aztec node), running the tx processor using a world-state synced to the last proven block, using the prover orchestrator to prove, and the l1 publisher to submit the proof back to L1.

Below is an example of a node and a prover node running side by side with the following commands:

```
ETHEREUM_HOST='http://localhost:8545' DEPLOY_AZTEC_CONTRACTS=true SEQ_SKIP_SUBMIT_PROOFS=1 SEQ_PUBLISHER_PRIVATE_KEY=0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d \
yarn aztec start --node --sequencer --prover --archiver --pxe
```

```
LOG_LEVEL=verbose ETHEREUM_HOST='http://localhost:8545' AZTEC_NODE_URL='http://localhost:8080' \
yarn aztec start -p 8090 --prover-node --prover --archiver
```

[Screencast from 2024-07-19 18-42-27.webm](https://github.com/user-attachments/assets/556edf0e-9843-481a-a488-1a497728194f)

See #7346

spalladino committed:
feat: p2p tx pool holds mined unproven txs

spalladino committed:
chore: Refactor factory and get-config methods

spalladino committed:
feat: Initial prover-node package

alexghr commented on yarn-project/prover-node/src/factory.ts:
```diff
@@ -0,0 +1,43 @@
+
+  const prover = await createProverClient(config, worldStateSynchronizer, archiver);
+
+  // REFACTOR: Move publisher out of sequencer package and into an L1-related package
+  const publisher = getL1Publisher(config);
```
❤️

alexghr commented on yarn-project/prover-node/src/job/block-proving-job.ts:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    this.state = 'awaiting-prover';
+    const { block, aggregationObject, proof } = await this.prover.finaliseBlock();
```
IIUC `startNewBlock` starts a block containing txs while `finaliseBlock` waits for the block rollup to complete? (sueprblock?)

spalladino replied:
For now everything is just block since we're still doing one proof per block. I'll probably change the API once we switch to batches.

alexghr commented on yarn-project/p2p/src/client/p2p_client.ts:
```diff
@@ -126,11 +133,12 @@ export class P2PClient implements P2P {
-    });
-    this.synchedBlockNumber = store.openSingleton('p2p_pool_last_l2_block');
+    const l2DownloaderOpts = { maxQueueSize: p2pL2QueueSize, pollIntervalMS: checkInterval };
+    this.latestBlockDownloader = new L2BlockDownloader(l2BlockSource, l2DownloaderOpts);
+    this.provenBlockDownloader = new L2BlockDownloader(l2BlockSource, { ...l2DownloaderOpts, proven: true });
```
I think this is fine for now but I'm thinking tha two downlaoder listening to the same source would mean the same block is read multiple times. 
I'd make the source & downlaoder event emitters/readable streams, but definitely sits in the realm of opimization/nice to haves and can be tackled independently of this PR.

spalladino replied:
Actually it doesn't: the L2Downloader, unlike its name describes, is just a queue that pulls from an Archiver. So each block is only downloaded and processed once.

spalladino replied:
Good point on the remote archiver, hadn't considered it! Agree in that we should revisit this component, and that we should think it so it properly handles reorgs (which is a pending assignment all across the board). But I'd push it for a future PR if that's ok.

alexghr reviewed (commented):
Looks good to me 👍

spalladino committed:
chore: Add prover-node e2e test

spalladino commented on yarn-project/aztec-node/src/aztec-node/config.ts:
```diff
@@ -14,23 +14,21 @@ import { fileURLToPath } from 'url';
     /** Whether the prover is disabled for this node. */
     disableProver: boolean;
-
-    /** A URL for an archiver service that the node will use. */
-    archiverUrl?: string;
```
This was moved to the ArchiverConfig

spalladino committed:
feat: Run prover-node via aztec start

Includes:
- Patched getInitialHeader
- Get proven block number from pxe
- Recurringly query unproven chain from prover-node

PhilWindle commented on yarn-project/aztec/src/cli/cli.ts:
```diff
@@ -32,7 +32,8 @@ export function injectAztecCommands(program: Command, userLog: LogFn, debugLogge
     .option('-a, --archiver [options]', cliTexts.archiver)
     .option('-s, --sequencer [options]', cliTexts.sequencer)
-    .option('-r, --prover [options]', cliTexts.prover)
+    .option('-r, --prover [options]', cliTexts.proverAgent)
```
I'm not entirely convinced of the way in which we are deviating from the current cli command system. If I want to run a prover node why wouldn't I just run:

`aztec start --node --prover`

Is it solely about being able to run multiple seperate prover nodes and only have a single transaction pool that they pull from? Is this likely? Is it even necessary to run multiple prover nodes?

spalladino replied:
This is mostly for backwards compatibility: we still support having a sequencer generate and publish proofs as it generates a block (this will be removed soon-ish), so we still need to handle the configuration of "sequencer node with local prover-agent", which is `--node --sequencer --prover`. Once we deprecate that, I agree on the need to revisit the CLI options here.

PhilWindle commented on yarn-project/sequencer-client/src/sequencer/sequencer.ts:
```diff
@@ -378,7 +386,15 @@ export class Sequencer {
-    return min >= this.lastPublishedBlock;
+    const [worldState, p2p, l2BlockSource, l1ToL2MessageSource] = syncedBlocks;
+    const result = min >= this.lastPublishedBlock;
+    this.log.debug(`Sync check to last published block ${this.lastPublishedBlock} ${result ? 'succeeded' : 'failed'}`, {
```
Doesn't this spam the logs? Even at debug, that would be pretty annoying.

spalladino replied:
A bit, but I'd argue that debug should be annoying, and it's useful for understanding the state of the components that act as dependencies to the sequencer. I'd keep it if possible.

PhilWindle reviewed (commented):
Some comments

spalladino commented on yarn-project/p2p/src/client/p2p_client.ts:
```diff
@@ -193,8 +207,10 @@ export class P2PClient implements P2P {
+    await this.latestBlockDownloader.stop();
+    await this.provenBlockDownloader.stop();
     this.log.debug('Stopped block downloader');
+    // TODO: This running promise will never resolve if there are no new blocks to sync.
```
Good call, added a timeout of 1s!

spalladino committed:
Address comments from PR
